rust/ip-packet/src/proptests.rs

```diff
@@ -0,0 +1,233 @@
+        })
+}
+
+fn udp_packet_v6() -> impl Strategy<Value = MutableIpPacket<'static>> {
+    (
+        any::<Ipv6Addr>(),
+        any::<Ipv6Addr>(),
+        any::<u16>(),
+        any::<u16>(),
+        any::<Vec<u8>>(),
+    )
+        .prop_map(|(src, dst, sport, dport, payload)| {
+            udp_packet(src.into(), dst.into(), sport, dport, payload)
+        })
+}
+
+fn icmp_packet_v4() -> impl Strategy<Value = MutableIpPacket<'static>> {
+    (
+        any::<Ipv4Addr>(),
+        any::<Ipv4Addr>(),
+        any::<u16>(),
+        any::<u16>(),
+        any::<IcmpKind>(),
+    )
+        .prop_map(|(src, dst, id, seq, kind)| icmp_packet(src.into(), dst.into(), id, seq, kind))
+}
+
+fn icmp_packet_v4_header_options() -> impl Strategy<Value = MutableIpPacket<'static>> {
+    (
+        any::<Ipv4Addr>(),
+        any::<Ipv4Addr>(),
+        any::<u16>(),
+        any::<u16>(),
+        any::<IcmpKind>(),
+        (5u8..15),
+    )
+        .prop_map(|(src, dst, id, seq, kind, header_length)| {
+            icmp4_packet_with_options(src.into(), dst.into(), id, seq, kind, header_length)
+        })
+}
+
+fn icmp_packet_v6() -> impl Strategy<Value = MutableIpPacket<'static>> {
+    (
+        any::<Ipv6Addr>(),
+        any::<Ipv6Addr>(),
+        any::<u16>(),
+        any::<u16>(),
+        any::<IcmpKind>(),
+    )
+        .prop_map(|(src, dst, id, seq, kind)| icmp_packet(src.into(), dst.into(), id, seq, kind))
+}
+
+fn packet() -> impl Strategy<Value = MutableIpPacket<'static>> {
+    prop_oneof![
+        tcp_packet_v4(),
+        tcp_packet_v6(),
+        udp_packet_v4(),
+        udp_packet_v6(),
+        icmp_packet_v4(),
+        icmp_packet_v6(),
+    ]
+}
+
+#[test_strategy::proptest()]
+fn can_translate_dst_packet_back_and_forth(
+    #[strategy(packet())] packet: MutableIpPacket<'static>,
+    #[strategy(any::<Ipv4Addr>())] src_v4: Ipv4Addr,
+    #[strategy(any::<Ipv6Addr>())] src_v6: Ipv6Addr,
+    #[strategy(any::<IpAddr>())] dst: IpAddr,
+) {
+    let original_source = packet.source();
+    let original_destination = packet.destination();
+    let original_packet = packet.packet().to_vec();
+
+    let original_source_v4 = if let IpAddr::V4(v4) = original_source {
+        v4
+    } else {
+        Ipv4Addr::UNSPECIFIED
+    };
+    let original_source_v6 = if let IpAddr::V6(v6) = original_source {
+        v6
+    } else {
+        Ipv6Addr::UNSPECIFIED
+    };
+
+    let packet = packet.translate_destination(src_v4, src_v6, dst).unwrap();
+
+    assert!(packet.source() == IpAddr::from(src_v4) || packet.source() == IpAddr::from(src_v6) || packet.source() == original_source, "either the translated packet was set to one of the sources or it wasn't translated and it kept the old source");
+    assert_eq!(packet.destination(), dst);
+
+    let mut packet = packet
+        .translate_destination(original_source_v4, original_source_v6, original_destination)
+        .unwrap();
+    packet.update_checksum();
```

Comment:
I'd expect the translation to give me valid checksums.

Reply:
the reason it's not like that is for performance, we want the `update_checksum` to be the very last thing that happens to a packet before writing it after all mangling.

It might be a bit of an early optimization but I'm not sure it's that harmful.

Reply:
Maybe we can express this in the type-system at some point so you can only go back to a regular `IpPacket` via a function that does the checksumming for you?

Reply:
that makes sense, though it's a bit of a more invasive check, let's save it for a future PR